MoltenVK/MoltenVK/Commands/MVKCommandPipelineStateFactoryShaderSource.h

```diff
@@ -168,5 +168,97 @@ kernel void cmdCopyBufferToImage3DDecompressTempBufferDXTn(constant uint8_t* src
     }                                                                                                           \n\
 }                                                                                                               \n\
                                                                                                                 \n\
+#if __METAL_VERSION__ == 210                                                                                    \n\
+// This structure is missing from the MSL headers. :/                                                           \n\
+struct MTLStageInRegionIndirectArguments {                                                                      \n\
+    uint32_t stageInOrigin[3];                                                                                  \n\
+    uint32_t stageInSize[3];                                                                                    \n\
+};                                                                                                              \n\
+#endif                                                                                                          \n\
+                                                                                                                \n\
+kernel void cmdDrawIndirectConvertBuffers(const device char* srcBuff [[buffer(0)]],                             \n\
+                                          device char* destBuff [[buffer(1)]],                                  \n\
+                                          constant uint32_t& srcStride [[buffer(2)]],                           \n\
+                                          constant uint32_t& inControlPointCount [[buffer(3)]],                 \n\
+                                          constant uint32_t& outControlPointCount [[buffer(4)]],                \n\
+                                          constant uint32_t& drawCount [[buffer(5)]],                           \n\
+                                          uint idx [[thread_position_in_grid]]) {                               \n\
+    if (idx >= drawCount) { return; }                                                                           \n\
+    const device auto& src = *reinterpret_cast<const device MTLDrawPrimitivesIndirectArguments*>(srcBuff + idx * srcStride);\n\
+    device char* dest = destBuff + idx * (sizeof(MTLDispatchThreadgroupsIndirectArguments) + sizeof(MTLDrawPatchIndirectArguments));\n\
+#if __METAL_VERSION__ >= 210                                                                                    \n\
```

Comment:
Curiouser and curiouser.

Just spent quite a while beating my head against the proverbial wall running this against published Dota2.

`vkCreateDevice()` causes `_MVKStaticCmdShaderSource` to strangely fail to compile with a bunch of:

```
program_source:160:101: error: unknown type name 'MTLDrawPatchIndirectArguments'; did you mean 'MTLDrawPrimitivesIndirectArguments'?
    device char* dest = destBuff + idx * (sizeof(MTLDispatchThreadgroupsIndirectArguments) + sizeof(MTLDrawPatchIndirectArguments));
                                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                                                                    MTLDrawPrimitivesIndirectArguments
```

It turns out that `MVKDevice::getMTLCompileOptions()` is unexpectedly returning a `MTLCompileOptions` instance whose `languageVersion` parameter is set to `MTLLanguageVersion1_1` instead of `MTLLanguageVersion2_1`.

This is probably related to a similar item you [speculated on in the DXVK issue](https://github.com/KhronosGroup/MoltenVK/issues/203#issuecomment-452868875).

So...a couple of thoughts:

1. `_MVKStaticCmdShaderSource` should adapt to MSL versions to conditionally elide MSL artifacts that are MSL version dependent.

2. `MVKDevice::getMTLCompileOptions()` should probably force the `languageVersion` to the latest MSL version available as determined by feature sets. We already track this in `MVKDevice` so it should be an easy change.

I've confirmed that (2) fixes the issue for Dota2...so I'll polish it up and do a separate PR for that. 

Can you take care of (1) as part of this PR...otherwise we might see legitimate crashes if people really run on macOS 10.11 / iOS 9.0?

Reply:
I've added a preprocessor directive to only compile in the tessellation helpers on Metal 1.2+.

Reply:
Great! Thanks.

I'll make the `languageVersion` changes I suggested on another PR.

One other topic of discussion:

When I started looking into this issue...I started wondering about whether we should have a config option / env var that enables tessellation...with it either enabled or disabled by default.

I was initially speculating that Dota2 might now be responding to the `VkPhysicalDeviceFeatures::tessellationShader` feature...and then making use of tessellation functionality that you haven't been able to support yet (or at all...eg isolines).

What do you think? Might some apps detect tessellation and start using it in ways that are not supported yet? Should we provide the option to deselect `VkPhysicalDeviceFeatures::tessellationShader` via an env var? 

And if so...should it be on or off by default?

Reply:
> What do you think? Might some apps detect tessellation and start using it in ways that are not supported yet?

For those that use Vulkan natively (e.g. Dota2, DOOM), this is certainly a possibility. But for D3D11 games (via DXVK), we were already faking this feature to make them run at all, so this should help them if anything.

> Should we provide the option to deselect `VkPhysicalDeviceFeatures::tessellationShader` via an env var?

It might be good to have one for now while we work out any remaining bugs. This will enable users to keep running games even if tessellation is broken.

> And if so...should it be on or off by default?

I don't know. I'm leaning towards leaving it on by default, since I want to know if there are bugs that I should go fix.